docs/language/sdk/services.mdx

```diff
@@ -0,0 +1,540 @@
+provider.
+
+Once the client has connected to a provider (with the help of the
+service manager), its main job consists of serializing the arguments
+and return values, and to map the method calls to the appropriate
+indexes.
+
+For example, the aforementioned `RandomGeneratorService` could be
+implemented as follows:
+
+```
+import system.services show ServiceClient ServiceSelector
+import .service
+
+class RandomGeneratorServiceClient extends ServiceClient:
+  static SELECTOR ::= RandomGeneratorService.SELECTOR
+
+  constructor selector=SELECTOR:
+    assert: selector.matches SELECTOR
+    super selector
+
+  generate limit/int -> int:
+    return invoke_ RandomGeneratorService.GENERATE_INDEX limit
+```
+
+The `RandomGeneratorServiceClient` class extends the `ServiceClient`
+base clase which implements the communication with the service manager.
+
+The `super` call takes a `ServiceSelector` object as argument. This
+object describes the service that the client wants to connect to. By
+default the selector from the `RandomGeneratorService` interface is
+used, but users can also provide their own selector as long as its
+compatible (the `assert` call checks this).
+
+The `generate` method then implements the actual service. It calls the
+`invoke_` method on the super class, passing the index of the method
+to call, and the arguments to the method. The `invoke_` method then
+serializes the arguments, sends them to the service provider, and
+waits for the response. It then deserializes the response and returns
+it.
+
+### Using a service client
+A client needs the a provider to be available. See how to
+[use providers](#using-a-service-provider) for more information.
+
+The common way to use a service is to construct it and to `open` it. The
+`open` call contacts the service manager and finds a matching provider.
+
+```
+main:
+  client := RandomGeneratorServiceClient
+  client.open
+  10.repeat:
+    print "random = $(client.generate 100)"
+```
+
+Often, the service can also be in a lazy-initialized constant:
+
+<!-- ANALYZE CODE -->
+```
+service_/RandomGeneratorServiceClient? ::= (RandomGeneratorServiceClient).open
+    --if_absent=: null
+```
+
+The `--if_absent` block is invoked when we cannot find the requested service.
+You can provide a timeout if you're willing to
+wait a bit for the service to appear:
+
+<!-- RESET CODE -->
+<!-- ANALYZE CODE -->
+<!-- HIDDEN CODE class RandomGeneratorServiceClient: -->
+<!-- HIDDEN CODE   open --timeout [--if_absent]: return null -->
+```
+service_/RandomGeneratorServiceClient? ::= (RandomGeneratorServiceClient).open
+    --timeout=(Duration --s=2)
+    --if_absent=: null
+```
+
+## Serialization
+As part of the communication between clients and providers, messages
+are serialized in an efficient way. Messagse must be either `null`,
+integers, booleans, floats, strings, byte arrays, lists or maps
+(with string keys). Any type that is not one of these must be converted
```

Comment:
Is the string keys really a requirement?

Reply:
no.
done.